### DISCUSSION TITLE
Controlling ioSender with Arduino Leonardo as Keyboard

### DISCUSSION THREAD

**Discussion by mintracer** [terjeio/ioSender#295] [Q&A] [answered]
Hi all!

well i would like to build a small easy working controlbox to control the simple commands as for example 

Program run - green Button - ALT+R
Pause - orange Button - Space
Stop - red Button - ALT-S 
 etc... 

The Arduino Leonardo can act as an HID and simulate a Keyboard...

Now my Problem, not all Shortcuts are working... even not with a real Keyboard... 

Do i need to activate them? 

I´m running Windows 11

Thx for helping me out! 

Cheers... :)

**Comment by terjeio** [comment]
> Now my Problem, not all Shortcuts are working... even not with a real Keyboard...

Which?

FYI you can write out the current mappings by clicking on _Save keymappings_ in the _Settings: App_ tab.

**Reply by mintracer** [reply]
all tested with real keyboard

don´t work -->

ALT + R  (grbl mainscreen)
Space
CRTL+H
CRTL+U

-------------------------------------------

works ->
ALT+R (Probing Tab)
CRTL+SHIFT (Axis XYZ) to zero
CRTL+SHIFT+0
CRTL+SHIFT+R 
CRTL++/-  (not on the NUMPAD)
CRTL+SHIFT++/-  (not on the NUMPAD)
all the Functionkeys for the macros (F1, F2 ...)

**Reply by terjeio** [reply]
All works for me in the _Grbl_ tab (Win10).
ALT + R requires program to be loaded.
Space requires program to be running.
CTRL + H requires homing enabled.
CTRL + U requires controller in a mode that requires unlock ("\<Unlock\> to continue." shown in status bar). Use SHIFT + CTRL + R to perform a soft reset first if in alarm mode.

**Reply by mintracer** [reply]
I tested the Shortcuts with a Job loaded, and running...

**Reply by terjeio** [reply]
Do you have another program running that claims the keys then?

**Reply by mintracer** [reply]
no, the strange thing is, that in the probing tab, ALT+R works... so when some App would claim the shortcut, it should work their either... 
maybe its Windows 11  - BTW. im running the latest original ioSender, nothing self compiled.

**Comment by pad73** [comment]
The same problem for me. Spacebar and other Ctrl+ and Alt+ shortcuts are not working. Only numpad selecting of distance and feed is working. I can use only cursor (mouse) and click. Im have win10.

**Reply by terjeio** [reply]
Could it be something wrong with my Win10 machine then?  Because the keys are working for me.
I'll try with another Win10 and a Win7 machine - maybe I can replicate the issue with them.

**Reply by pad73** [reply]
Great! Shortcuts are very important. Nobody want to hunt small buttons with their mouse :) And with shortcuts you can use external gamepad easily (for most needed instructions).

**Reply by mintracer** [reply]
what i thought about was, that maybe the ASCII are different depending on the Keyboard Layout...

**Reply by terjeio** [reply]
> what i thought about was, that maybe the ASCII are different depending on the Keyboard Layout...

I am using [keycodes](https://github.com/terjeio/ioSender/blob/b6b66e44f9e82b76a2e19dfdf7814c7b20ef8319/CNC%20Controls/CNC%20Controls/JobControl.xaml.cs#L153-L175) not the characters produced so the layout should not matter.

Perhaps I should add a startup option that logs the keycodes to console?

**Reply by mintracer** [reply]
this would be great for testing / debugging purposes...

**Reply by terjeio** [reply, marked as answer]
I've uploaded a special edge version, [ioSender Edge Keytest.zip](http://www.io-engineering.com/downloads), that logs keypresses to the console when _Verbose_ is checked.
Note that since logging is done in keypress event handlers it might interfere with normal operation.
Examples:
Pressing space for feed hold:
```
<Run|MPos:66.018,91.805,-13.770|Bf:9,1023|FS:662,500>
Space None None False
<Run|MPos:67.768,92.068,-13.770|Bf:14,1023|FS:292,500>
Space None None True
<Hold:1|MPos:67.783,91.148,-13.770|Bf:18,1023|FS:0,500>
```
There are two events logged, keydown and keyup. The keyup log above has _true_ added to it since the key was claimed as a shortcut.
Note that "printable" characters like Space will not be claimed as a shortcut if an input field has focus. E.g. if the MDI has focus, click in the _Program_, _3D View_ or _Console_ tab to move focus away.

Pressing Alt + R for cycle start (with gcode loaded):
```
System R Alt False
LeftAlt None None False
System R None True
<Idle|MPos:64.560,89.298,-13.770|Bf:35,231|FS:0,500|A:CF>
[GC:G1 G54 G92 G17 G21 G90 G94 G49 G98 G50 M4 M8 T0 F100 S500.]
<Idle|MPos:64.560,89.298,-13.770|Bf:35,231|FS:0,500|WCO:61.260,85.768,-14.840>
[GC:G1 G54 G92 G17 G21 G90 G94 G49 G98 G50 M4 M8 T0 F100 S500.]
<Idle|MPos:64.560,89.298,-13.770|Bf:35,231|FS:0,500>
<Run|MPos:64.508,89.243,-13.770|Bf:0,435|FS:788,500>
[GC:G2 G54 G92 G17 G21 G90 G94 G49 G98 G50 M4 M8 T0 F1500 S500.]
```

Do _not_ use this version for regular work!

**Reply by mintracer** [reply]
i will test it... thank you for the support!!

**Reply by pad73** [reply]
I can test today evening.

**Reply by pad73** [reply]
Great, this "keyversion" works well. I tested basic functions only ... Home, Start, Hold .... other keys which I tested before, it means switching feed, distance, are also OK.
Only Start (Alt+R) makes not only Start but also some Windows sound. I don´t know why. IMHO you can think if is possible to put Start Cycle and Feed Hold to only one key (Spacebar). Even situation you can only start program or only pause (feed) program. So big key seems to be ideal than Alt+R. But nevermind, it is important it works now. Thank you.

Is it possible to make own shortcuts? Not with macro function. Only to change "letter" definition of some commands

**Reply by terjeio** [reply]
> Only Start (Alt+R) makes not only Start but also some Windows sound. I don´t know why.

I do not know either. Have to ask Microsoft?

> IMHO you can think if is possible to put Start Cycle and Feed Hold to only one key (Spacebar).

I do not want that, cycle start should be a two-key combination for safety reasons.

> Is it possible to make own shortcuts? Not with macro function. Only to change "letter" definition of some commands

Yes, write the shortcuts to a file by clicking on _Save keymappings_ in the _Settings: App_ tab and [edit it](https://github.com/terjeio/ioSender/discussions/315#discussioncomment-6484476) to match your needs.

**Reply by pad73** [reply]
Terjeio, you are the boss 👍. I understand your answers. About start / hold with only one key it was just idea. I must try shortcuts, I thought the button "save shortcuts" is only to save, I didn't expected it is to open some choices.
I must learn your iosender better, I'm not beginner in milling, but I'm beginner with iosender. I can see choice "highlight completed cut" but it is inactive. Highlight realized data on 3D wiev is also very usable.
